com.unity.ml-agents/Runtime/InplaceArray.cs

```diff
@@ -0,0 +1,250 @@
+using System;
+using System.Collections.Generic;
+
+namespace Unity.MLAgents
+{
+    /// <summary>
+    /// An array-like object that stores up to four elements.
+    /// This is a value type that does not allocate any additional memory.
+    /// </summary>
+    /// <remarks>
+    /// This does not implement any interfaces such as IList, in order to avoid any accidental boxing allocations.
+    /// </remarks>
+    /// <typeparam name="T"></typeparam>
+    public struct InplaceArray<T> where T : struct
```

Comment:
I would add a 
```
public struct ObservationShape : InplaceArray<int> { }
```
And same for DimensionProperties. Thoughts? Maybe overkill

Reply:
I think the constructors wouldn't work without adding them to ObservationShape too (plus DimensionProperties), which I'd rather not do.

And `using ObservationShape = InplaceArray<int>;` is only scoped for the file it's declared in, so wouldn't help any calling code.